### PR TITLE
loongarch: add glibc start.S to make zig-bootstrap work for loongarch64-linux-gnu

### PR DESCRIPTION
**Pull request by yxd-ym**
`start.S` and `regdef.h` are from glibc 2.40.

Current LLVM 18 does not support `.cfi_label` yet in `start.S`.

We need this in LLVM to make this code work in zig-bootstrap.
https://github.com/llvm/llvm-project/pull/97922

Also, I'm not sure whether I should put the `regdef.h` file the path here in this patch or should put it at `lib/libc/include/loongarch64-linux-gnu/sys`

### CODE REVIEW

**Comment by alexrp** [comment]
That's odd; I definitely updated the start files in #20993. Hmm...

**Comment by alexrp** [comment]
> We need this in LLVM to make this code work in zig-bootstrap.
> [llvm/llvm-project#97922](https://github.com/llvm/llvm-project/pull/97922)

This fix is in LLVM 19; we also need it for RISC-V.

**Comment by alexrp** [comment]
Ok, I see what went wrong. The [update procedure](https://github.com/ziglang/zig/wiki/Updating-libc#glibc) doesn't document the fact that start files for new architectures need to be added by hand.

However, I believe we might need some more headers:

```
❯ fd '\.h' lib/libc/glibc/**/riscv
lib/libc/glibc/sysdeps/riscv/bits/endianness.h
lib/libc/glibc/sysdeps/riscv/nptl/bits/pthreadtypes-arch.h
lib/libc/glibc/sysdeps/unix/sysv/linux/riscv/kernel-features.h
lib/libc/glibc/sysdeps/unix/sysv/linux/riscv/sysdep.h
```

It's not actually clear to me why we're duplicating these between `lib/libc/include/...` and `lib/libc/glibc/...`, but let's just do it consistently at least.

> Also, I'm not sure whether I should put the `regdef.h` file the path here in this patch or should put it at `lib/libc/include/loongarch64-linux-gnu/sys`

I *think* you placed it correctly, but just to be sure, I'll build glibc for `loongarch64-linux-gnu` again locally and get back to you on this.

**Comment by alexrp** [comment]
> I _think_ you placed it correctly, but just to be sure, I'll build glibc for `loongarch64-linux-gnu` again locally and get back to you on this.

Yes, it's correct; it doesn't get installed to `/usr/include`.

File: src/glibc.zig [hunk 1]
```diff
@@ -649,6 +651,9 @@ fn add_include_dirs_arch(
             try args.append("-I");
             try args.append(try path.join(arena, &[_][]const u8{ dir, "riscv" }));
         }
+    } else if (arch.isLoongArch()) {
+        try args.append("-I");
+        try args.append(try path.join(arena, &[_][]const u8{ dir, "loongarch" }));
```
Comment (alexrp) [inline src/glibc.zig:656]: Note that, after adding the extra headers I mentioned, you'll need to add the `nptl` subdirectory here like it's done for RISC-V above.

Comment (yxd-ym) [reply src/glibc.zig]: done

File: lib/libc/glibc/sysdeps/loongarch/nptl/bits/pthreaddef.h [hunk 2]
```diff
@@ -0,0 +1,32 @@
+/* pthread machine parameter definitions.
```
Comment (alexrp) [inline lib/libc/glibc/sysdeps/loongarch/nptl/bits/pthreaddef.h:1]: Oh, I didn't realize LoongArch doesn't have a `pthreadtypes-arch.h`. I don't think we need this header then; they're not the same. So the only header we were actually missing was `bits/endianness.h`. So I guess just remove this one after all, and 68451422d16268cd6acc1b629ce60fe4a3a9d7c7 won't be needed then either.

Sorry for the confusion. :slightly_smiling_face:

Comment (yxd-ym) [reply lib/libc/glibc/sysdeps/loongarch/nptl/bits/pthreaddef.h]: Removed. Is there any specific rule to tell whether a arch-specific header file is needed?

Comment (alexrp) [reply lib/libc/glibc/sysdeps/loongarch/nptl/bits/pthreaddef.h]: Not that I'm aware of. Like I mentioned above, I'm not even sure *why* we duplicate these headers in `lib/libc/glibc/...` when they (usually) also end up in `lib/libc/include/...`. We should probably clean this up at some point.

For now, I just recommended duplicating the ones that were duplicated for RISC-V and also exist for LoongArch, which... just comes down to `bits/endianness.h`, actually.

**Review by alexrp — APPROVED** [review]

**Pull request merged**